diff --git a/modules/janus/lib/REST/Methods.php b/modules/janus/lib/REST/Methods.php
index de9ed55..c485688 100644
--- a/modules/janus/lib/REST/Methods.php
+++ b/modules/janus/lib/REST/Methods.php
@@ -474,7 +474,12 @@ class sspmod_janus_REST_Methods
                 $result[$meta->getKey()] = $meta->getValue();
             }
         }
-
+
+        // Add disable consent
+        foreach(array_keys($entityController->getDisableConsent()) as $entityIndex => $entityUrl) {
+            $result['disableConsent:' . $entityIndex] = $entityUrl;
+        }
+
         return $result;
     }
