diff --git a/modules/janus/UPGRADE b/modules/janus/UPGRADE
index 841a81f..e537670 100644
--- a/modules/janus/www/dashboard.php
+++ b/modules/janus/www/dashboard.php
@@ -90,7 +90,7 @@ if(isset($_POST['add_usersubmit'])) {
 if(isset($_POST['submit'])) {
     $selectedtab = '1';
     if (!empty($_POST['entityid'])) {
-        if (check_uri($_POST['entityid'])) {
+        if (check_uri($_POST['entityid']) || true) {  # HACK: For SURFnetGuests : https://espee-test.surfnet.nl/federate/metadata/saml20/SURFnetGuests
             if(!isset($_POST['entityid']) || empty($_POST['entitytype'])) {
                 $msg = 'error_no_type';
                 $old_entityid = $_POST['entityid'];
diff --git a/modules/janus/UPGRADE b/modules/janus/UPGRADE
index 841a81f..e537670 100644
--- a/modules/janus/www/editentity.php
+++ b/modules/janus/www/editentity.php
@@ -135,7 +135,7 @@
 
     // Change entityID
     if(isset($_POST['entityid']) && $guard->hasPermission('changeentityid', $entity->getWorkflow(), $user->getType())) {
-        if(check_uri($_POST['entityid'])) {
+        if(check_uri($_POST['entityid']) || true) { # HACK: For SURFnetGuests : https://espee-test.surfnet.nl/federate/metadata/saml20/SURFnetGuests
             $entityIdNeedsUpdating = $_POST['entityid'] != $entity->getEntityid();
             if($entityIdNeedsUpdating) {
                 $userController = new sspmod_janus_UserController($janus_config);
